src/diffusers/models/modeling_utils.py

```diff
@@ -314,17 +328,64 @@ def save_pretrained(
         state_dict = model_to_save.state_dict()
 
         weights_name = SAFETENSORS_WEIGHTS_NAME if safe_serialization else WEIGHTS_NAME
-        weights_name = _add_variant(weights_name, variant)
+        weights_name = _add_variant(weights_name, variant, add_suffix_keyword=True)
 
         # Save the model
-        if safe_serialization:
-            safetensors.torch.save_file(
-                state_dict, Path(save_directory, weights_name).as_posix(), metadata={"format": "pt"}
-            )
-        else:
-            torch.save(state_dict, Path(save_directory, weights_name).as_posix())
+        state_dict_split = split_torch_state_dict_into_shards(
+            state_dict, max_shard_size=max_shard_size, filename_pattern=weights_name
+        )
+        # Clean the folder from a previous save
+        for filename in os.listdir(save_directory):
+            full_filename = os.path.join(save_directory, filename)
+            # If we have a shard file that is not going to be replaced, we delete it, but only from the main process
+            # in distributed settings to avoid race conditions.
+            weights_no_suffix = weights_name.replace(".bin", "").replace(".safetensors", "")
+
+            # make sure that file to be deleted matches format of sharded file, e.g. pytorch_model-00001-of-00005
+            filename_no_suffix = filename.replace(".bin", "").replace(".safetensors", "")
+            reg = re.compile(r"(.*?)-\d{5}-of-\d{5}")
+
+            if (
+                filename.startswith(weights_no_suffix)
+                and os.path.isfile(full_filename)
+                and filename not in state_dict_split.filename_to_tensors.keys()
+                and is_main_process
+                and reg.fullmatch(filename_no_suffix) is not None
+            ):
+                os.remove(full_filename)
 
-        logger.info(f"Model weights saved in {Path(save_directory, weights_name).as_posix()}")
+        # Save the model
+        for filename, tensors in state_dict_split.filename_to_tensors.items():
+            shard = {tensor: state_dict[tensor] for tensor in tensors}
+            if safe_serialization:
+                # At some point we will need to deal better with save_function (used for TPU and other distributed
+                # joyfulness), but for now this enough.
+                safetensors.torch.save_file(shard, os.path.join(save_directory, filename), metadata={"format": "pt"})
+            else:
+                torch.save(shard, Path(save_directory, filename).as_posix())
```

Comment:
Using `Path(save_directory, filename).as_posix()` will convert the path to a "unix-like string", meaning it would fail on Windows -or at least not result in what we want-. Also I'm all for using `Path` more often but seems like here it would be the only occurrence of the method so I'd better stick to a `filepath = os.path.join(save_directory, filename)` for consistency.

```suggestion
            filepath = os.path.join(save_directory, filename)
            if safe_serialization:
                # At some point we will need to deal better with save_function (used for TPU and other distributed
                # joyfulness), but for now this enough.
                safetensors.torch.save_file(shard, filepath, metadata={"format": "pt"})
            else:
                torch.save(shard, filepath)
```

Reply:
We have https://github.com/huggingface/diffusers/pull/6564. But for consistency, I changed the `save_file` path to use `Path` as well.